### ISSUE TITLE
[D2A] Aufbau der D2A-Repositories

### ISSUE THREAD

**Issue by AKirsch1** [Programmiermethoden/Dungeon#621]
*Anmerkung: Basierend auf [Kommentar von @AMatutat](https://github.com/Programmiermethoden/Dungeon/issues/434#issuecomment-1532679532).*

![Repository Struktur](https://user-images.githubusercontent.com/32961997/236833681-507dac5b-7414-4c8c-a7f3-a4403e7c594e.png)

*Anmerkung: Die Namen der Repositories sind nicht final.*

*Notiz: Für Informationen zu git filter-repo siehe [hier](https://www.mankier.com/1/git-filter-repo).*

|Repository|Erläuterung|
|--|--|
|D2A|Enthält die GitHub Action, die über den Workflow ab dem Template-Repository in jedes darunter liegende Repository eingebunden wird.|
|Template|Enthält die Vorlagen für alle Aufgaben.|
|Tasks|(Optional) Enthält die Vorlagen für ausschließlich die Aufgaben, die die Studierenden bearbeiten sollen.|
|StudiSolution|Lösungen der Studierenden zu den Aufgaben.|
|HomeworkSolution|(Option 1): Die Lösungen werden wie bisher im gleichen Repository gesammelt wie die Vorlagen. Das Template-Repository ist ein über Subtree erstelltes Repository, welches nur die Vorlagen enthält.|
|SampleSolution|(Option 2): Enthält wie Option 1 die Lösungen zu den Vorlagen. Verhält sich im Gegensatz dazu aber wie StudiSolution.|

**Gegenüberstellung HomeworkSolution vs SampleSolution**

|HomeworkSolution|SampleSolution|
|--|--|
|- Neue Aufgaben können in ein und demselben Repository erstellt werden (Keine Arbeit in zwei Repositories gleichzeitig oder Rüberkopieren von Dateien notwendig).|- Entspricht dem Workflow der Studierenden.|
|- Kein doppelter Vorlagen-Code in zwei unterschiedlichen Repositories.|- Erlaubt das Testen durch D2A, ohne eine Unterscheidung zu StudiSolution machen zu müssen.|
|- Bereits Pull-Request mit dieser Lösung vorhanden.|- Ermöglicht das Arbeiten in den Vorgaben (größerer möglicher Aufgabenpool).|
||- Keine unterschiedlichen Build-Skripte und kein Ändern der Gradle-Konfiguration durch Source-Code, der an unterschiedlichen Stellen liegt.|

*Zusammenfassung:* Der Ansatz *SampleSolution* erscheint mir deutlich simpler und ermöglicht zusätzliche Aufgabentypen und Testmöglichkeiten. *HomeworkSolution* vereinfacht dagegen die Erstellung von neuen Aufgaben und wir haben bereits einen PR dazu.

**Erläuterung Task-Repository**

Studierenden wird pro Praktikumseinheit eine ausgewählte Liste an Aufgaben gestellt. Da im Template-Repository aber alle Vorlagen vorhanden sind, wird ein "Zwischenrepository" benötigt, welches Aufgaben zu Praktikumseinheiten zuordnet, die an Studierende verteilt werden. *Möglicher Aufbau des Repos: Zum Beispiel pro Praktikumseinheit ein Branch oder Aktualisieren bei jedem Aufgabenbeginn oder Praktikumseinheiten aufgeteilt in Unterordner.*

*Anmerkung: Die Erläuterung basiert darauf, dass ich nicht weiß, wie die Aufgaben vorher an Studierende gegeben wurden.*

*Anmerkung 2: @AMatutat hat mir mittlerweile gesagt, dass ihr im letzten Semester den Studis alle Aufgaben auf einmal zur Verfügung gestellt habt.*

**Comment by cagix** [comment]
@AKirsch1 Gibt es signifikate Vorteile, wenn jeder Task in einem eigenen Repo lebt (also bei "SampleSolution")? 

Ich habe den Maintenance-Aspekt und das Erstellen neuer Aufgaben im Blick. Da finde ich ein zentrales Repo deutlich angenehmer (also "_HomeworkSolution_"), weil ich da alles an einer Stelle habe und die zentrale Konfig auch nur einmal machen muss und mit einem `git subtree push` (oder `filter-tree`) den passenden Unterordner für die Studis "rausschieben" kann. 

Die "_SampleSolution_" hat für mich zudem den Nachteil, dass ich bei einer Überarbeitung der Vorgaben (Schnittstellen, Tests) diese erstmal "blind" im Vorgaberepo ("Template") bearbeiten würde, dann meine SampleSolution rebasen würde und schauen würde, ob alles passt. (Sicher kriegt man das auch noch besser hin mit Hilfsbranches im SampleSolution und Cherry-Pick und so, aber trotzdem ist das gefühlt ein recht aufwändiger Stunt.)

**Comment by cagix** [comment]
> > Gibt es signifikate Vorteile, wenn jeder Task in einem eigenen Repo lebt (also bei "SampleSolution")?
> Ich bin mir nicht sicher, ob ich das richtig verstehe, weil in "SampleSolution" eigentlich die Lösungen für alle Tasks drin sein sollen. Ansonsten hätte ich doch Unmengen an Repositories.

Ja, okay, da habe ich Dich scheinbar falsch verstanden. Ich hatte mitgenommen, dass bei der Variante "SampleSolution" pro Task ein eigenes Repo angelegt wird.

**Comment by cagix** [comment]
> Ich denke aber auch, dass man recht schnell zwischen den Varianten wechseln kann. Die "SampleSolution"-Variante braucht, wenn ich keinen Denkfehler habe, keine Anpassung des D2A: Wenn es in den "StudiSolution"s klappt, dann funktioniert auch "SampleSolution". Und auf der anderen Seite haben wir einen fertigen PR für die "HomeworkSolution"-Variante, bei dem nur noch wenige Anpassungen notwendig sein sollten. Also so viel Aufwand sollte es nicht sein, dass man beides mal ausprobieren kann, wenn alles steht. 

Bin grad nicht sicher, ob ich Dir folgen kann. In der einen Variante habe ich ein zentrales Repo mit dem Vorgabecode und den Lösungen, aus dem ich am Anfang des Semesters Ordner für die Studis rauslöse. In der anderen Variante habe ich ein Repo mit dem Vorgabecode und ein zweites mit den Lösungen. 

In der zweiten Variante arbeite ich also genau wie die Studis. Das hat den Vorteil, dass ich alles aus deren Sicht erlebe. Der Nachteil ist aber, dass ich bei der Pflege der Vorgaben einen ziemlichen Aufriss machen muss - zumindest habe ich bisher keine "einfache" Lösung. Es läuft immer auf "Zauber" mit komplexeren Branches oder Kopieren von Code o.ä. hinaus. Wenn Du hier mit einer halbwegs einfachen und stabilen Vorgehensweise um die Ecke kommst, bin ich sofort dabei :)

**Comment by AKirsch1** [comment]
> In der einen Variante habe ich ein zentrales Repo mit dem Vorgabecode und den Lösungen, aus dem ich am Anfang des Semesters Ordner für die Studis rauslöse. In der anderen Variante habe ich ein Repo mit dem Vorgabecode und ein zweites mit den Lösungen.

Genau so ist es von mir gedacht. Das mit dem "Rauslösen" wäre aber in beiden Varianten der Fall bzw. möglich (Das Tasks-Repo habe ich in der Tabelle ja als optional markiert).

> Wenn Du hier mit einer halbwegs einfachen und stabilen Vorgehensweise um die Ecke kommst, bin ich sofort dabei :)

Bisher habe ich noch keine Lösung dafür, aber ich werde mal schauen, ob ich da was finde.

**Comment by cagix** [comment]
git worktree ist ja eigentlich nur ein in einem zweiten ordner ausgecheckter branch. das klappt echt gut, für die ide ist das einfach ein weiteres projekt.

aber es ist halt nur ein (weiterer) branch im repo, der dann auch auf github landen würde (ok, vielleicht in einem zweiten remote?). wie würde das helfen im obigen szenario?
